File: src/chipschallenge/blockreactions/RedKeyTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Block.Type;
import chipschallenge.Game;
import chipschallenge.Inventory.Key;

/**
 *
 * @author patrik
 */
public class RedKeyTo extends BlockReaction {

    private RedKeyTo() {
    }
    private static RedKeyTo mInstance = null;

    public static synchronized RedKeyTo getInstance() {
        if (mInstance == null) {
            mInstance = new RedKeyTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        if (moving.isChip()) {
            takeKey(Key.RED);
            standing.replace(createBlock(Type.FLOOR));
        }
    }

    public boolean canMove(Block moving, Block standing) {
        return true;
    }
}

File: ChipsChallenge.Shared/Blockreactions/RedKeyTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Type = Block.Type;
    using Key = Inventory.Key;

    public class RedKeyTo : NoSlipReaction
    {
        private RedKeyTo()
        {
        }
        private static RedKeyTo instance;

        public static RedKeyTo Instance
        {
            get
            {
                lock (typeof(RedKeyTo))
                {
                    if (instance == null)
                    {
                        instance = new RedKeyTo();
                    }
                    return instance;
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            if (moving.Chip)
            {
                TakeKey(Key.RED);
                standing.Replace(CreateBlock(Type.FLOOR));
            }
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}